pkg/diagnostics/log/log.go

```diff
@@ -0,0 +1,280 @@
+package log
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	ct "github.com/daviddengcn/go-colortext"
+	"io"
+	"io/ioutil"
+	"runtime"
+	"strings"
+	"text/template"
+
+	"github.com/openshift/origin/pkg/version"
+)
+
+type LoggerOptions struct {
+	Level  int
+	Format string
+	Out    io.Writer
+}
+
+func (o *LoggerOptions) NewLogger() (*Logger, error) {
+	out := o.Out
+	if out == nil {
+		out = ioutil.Discard
+	}
+
+	ret, err := NewLogger(o.Level, o.Format, out)
+	return ret, err
+}
+
+type Level struct {
+	Level  int
+	Name   string
+	Prefix string
+	Color  ct.Color
+	Bright bool
+}
+
+func (l Level) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + l.Name + `"`), nil
+}
+
+func (l Level) MarshalYAML() (interface{}, error) {
+	return l.Name, nil
+}
+
+type Logger struct {
+	loggerType
+	level        Level
+	warningsSeen int
+	errorsSeen   int
+}
+
+// Internal type to deal with different log formats
+type loggerType interface {
+	Write(Entry)
+	Finish()
+}
+
+func NewLogger(setLevel int, setFormat string, out io.Writer) (*Logger, error) {
```

Comment:
I would want to use templates even if we didn't do json and yaml. I don't understand the objection to them. They are the only sane way to write some things. Taking them out doesn't even simplify anything AFAICS.